README
Math::GSL

Math::GSL is an interface to the GNU Scientific Library, using SWIG.  The GNU
Scientific Library (GSL) is a numerical library for C and C++ programmers. It
is free software under the GNU General Public License.  Math::GSL uses SWIG to
generate Perl bindings to *most* GSL functionality.

Currently Math::GSL requires at least Perl 5.8 and GSL 1.8 to compile. SWIG
is only needed if you want help work on Math::GSL and checkout code from the
Math::GSL git repository.

INSTALLATION

To install this module, run the following commands:

	perl Build.PL
	./Build
	./Build test
	./Build install clean

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Math::GSL

You can also look for information at:

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Math::GSL

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Math::GSL

    CPAN Ratings
        http://cpanratings.perl.org/d/Math::GSL

    Search CPAN
        http://search.cpan.org/dist/Math::GSL


COPYRIGHT AND LICENCE

Copyright (C) 2008 Jonathan Leto and Thierry Moisan

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.



Changes
=head1 Math::GSL Changes

=head1 v0.42 - June 2020
    - Disabled two tests in SF.t and GSL.t that failed on some platforms.

=head1 v0.41 - June 2020
    - Added support for GSL 2.6 (hakonhagland)
    - Added docker file support for building a CPAN distribution
    - Added support for gsl_odeiv_system (hakonhagland)
    - Add support for Debian-specific version 1.16.1, which caused test failures

=head1 v0.40 - December 2018
    - added support for GSL 2.5 (djerius)
    - vast improvements to Travis CI test infrastructure (djerius)
    - added random method in Math::GSL::Matrix (Rui Meira)
    - added Math::GSL::Multiset (standard gsl API and Perl OO)
    - gsl_linalg_givens now works
    - gsl_vector_const_subvector is now exported properly
    - Support for Sparse Matrices was added
    - Documentation for Math::GSL::Rstat, Math::GSL::SparseMatrix,
      Math::GSL::Diff
    - Math::GSL::Multifit now requires GSL 2.1 and higher
    - Math::GSL::Rstat now requires GSL 2.1 and higher

=head1 v0.39 - September 2016
    - Fix Debian Bug reported by Niko Tyni
        https://bugs.debian.org/826009
    - Extensive fixes to allow GSL 1.15 and GSL 1.16
      to continue to work with GSL 2.x
    - GSL 1.11 - 1.14 are no longer supported

=head1 v0.38 - September 2016
    - Fixes for compiling against various GSL
      versions (thanks CPANtesters!)

=head1 v0.37 - September 2016
    - Support for GSL 2.0, 2.1 and 2.2
    - Multilarge subsystem added for GSL >= 2.1

=head1 v0.36 - Fri Mar 25 23:32:03 PDT 2016
    - Fixed error messages on Matrix set_col and set_row
      methods (Jon Schutz)
    - Implemented array-based Bessel functions:
               gsl_sf_bessel_Jn_array
               gsl_sf_bessel_Yn_array
               gsl_sf_bessel_In_array
               gsl_sf_bessel_In_scaled_array
               gsl_sf_bessel_Kn_array
               gsl_sf_bessel_Kn_scaled_array
               gsl_sf_bessel_jl_array
               gsl_sf_bessel_jl_steed_array
               gsl_sf_bessel_yl_array
               gsl_sf_bessel_il_scaled_array

=head1 v0.35 - April 2015

    - Remotion of trailing whitespaces (Shlomi Fish)
    - Documentation fixes (Shlomi Fish)
    - Remotion of PkgConfig from runtime requirements
      (Bartosz Jakubski)
    - Indentation fixes (Bartosz Jakubski)
    - Installation feedback improvements (Bartosz Jakubski)

=head1 v0.34 - December 2014

    - Fixed bug on uncleared error variable (5.8.9)
    - Added OO transpose method for matrices (Ivan Baidakou)

=head1 v0.33 - September 2014
    - Added OO code for RNG->shuffle, RNG->choose and RNG->sample.
    - Complain when opening (fopen) a file from a tainted string.
    - Complex now has operator overloading for +,-, ==, !=, *, /
    - Added write/load methods to save a matrix to file.

=head1 v0.32 - June 2014
    - Correctly split the compiler name.
    - Added OO code to the QRNG module.
    - Fixed QRNG code that was failing on some machines.

=head1 v0.31 - June 2014
    - Fixed Matrix multiplication that was forcing square matrices.

=head1 v0.30 - June 2014
    - Fixed generated C files with wrong GSL version numbers.
    - Refactored Build.PL in order to use either gsl-config or PkgConfig.
    - Change Build.PL to return 0 when GSL library not found.
    - Use PkgConfig instead of ExtUtils::PkgConfig.
    - Math::GSL::Matrix constructor from a GSL matrix.
    - Reorganization on the Math::GSL::Matrix documentation.
	- Add symbols introduced on gsl 1.15.
	- Fixed in the correct way the prove/TAP::Harness problem.

=head1 v0.29 - June 2014
	- Fix issue when prove/TAP::Harness uses lib/ modules instead of blib/ ones.
    - Add functions introduced in gsl 1.16.
    - Fix issue when compiling on some Linuxes.
	- Math::GSL ->each method.
    - Math::GSL ->max and ->min methods.
	- Math::GSL ->dim method.
    - Math::GSL ->set_elem and ->get_elem methods.
    - Math::GSL concat_vertically and concat_horizontally.

=head1 v0.28 - Jan 2014
    - Fix build errors with clang
    - Improve POD to play nice with Perl 5.18
    - Teach Math::GSL about new functions added in GSL 1.16
    - Added 1D minimization tests

=head1 v0.27 - July 2012
    - Mathieu functions are now accessible
    - Improved detection of gsl
    - Fix bug in calling ->get() method on Math::GSL::Vectors
    - Greatly improved HTML formatting of documentation

=head1 v0.26 - September 2011
    - The Math::GSL::Min and Math::GSL::Roots are now functional - Yury Zavarin
    - Dirichlet random distribution support and greatly increased docs by Tom Nishimura <tnish@fastmail.jp>
    - gsl_version() function which returns the version of GSL being used by Tom Nishimura <tnish@fastmail.jp>
    - Fix '-bundle not allowed with -dynamiclib' build error on Mac OS X - Yury Zavarin
    - Build swig wrappers for different versions of GSL (1.11 - 1.15) - Yury Zavarin
    - Minimum SWIG version is now 2.x (only required for developers using the git repo)
    - Removed deprecated stuff from Build.PL and added Module::Build to configure_requires
    - Many small documentation improvements
    - Fix build on Windows - Sisyphus

=head1 v0.25 - April 2011
    - TODO RT#66882 properly so that a stable release with a passing test suite can be used.

=head1 v0.24 - April 2011
    - Fix https://rt.cpan.org/Ticket/Display.html?id=66882 (Tests fail on Perl's with DUSELONGDOUBLE)

=head1 v0.23 - March 2011
    - Fixed the tolerance of some tests that were causing CPANtesters failures
    - Fixed some doc typos

=head1 v0.22 - June 2010
    - Improved tests
    - == and != operators for Vectors
    - Fixed the export of hypergeometric functions (Thierry Moisan)
    - In instantiation of GSLBuilder in Build.PL, the value for the parameter
      include_dirs were corrected to an anonymous empty array, instead of an
      empty string.  (Jakob Ilves)
    - Operator overloading of vector equality (Thierry Moisan)
    - Various doc fixes (Pjotr Prins)
    - Typemap fixes (Vincent Danjean)


Build.PL
#!/usr/bin/perl -w
# This has been heavily modified from the Device::Cdio Build.PL
# Jonathan "Duke" Leto <jonathan@leto.net>

#    Copyright (C) 2006 Rocky Bernstein <rocky@cpan.org>
#
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 2 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
use strict;
use warnings;
use Config;
use Data::Dumper;
use Module::Build;
use lib 'inc';
use GSLBuilder;
use Ver2Func;
use File::Spec::Functions qw/:ALL/;
use Alien::GSL;

our $MIN_GSL_VERSION = "1.15";

my $gsl_conf;

check_alien_gsl_share_dir();
$gsl_conf = check_gsl_version();
$gsl_conf = check_gsl_version_pkgconfig() if not $gsl_conf;

my ($major,$minor,$tiny) = split /\./, $gsl_conf->{gsl_version};
my $ccflags = $gsl_conf->{gsl_cflags};
my $ldflags = $gsl_conf->{gsl_libs};

# In case GSL in installed in the system-wide directory, $ccflags is
# empty (because pkg-config remove -I/usr/include), but swig needs it

my $swig_flags = '';
$swig_flags = ' -I/usr/include' if $ccflags =~ /^\s*$/;

$ccflags .= " $Config{ccflags}"; # we need this for some Linuxes

## Swig produces a number of GCC warnings. Turn them off if we can.
$ccflags .= try_cflags("-Wall");
$ccflags .= try_cflags("-Wno-sometimes-uninitialized");
#$ccflags .= try_cflags("-Wno-strict-aliasing");
$ccflags .= try_cflags("-Wno-unused-function");
$ccflags .= try_cflags("-Wno-unused-value");
$ccflags .= try_cflags("-Wno-unused-function");
$ccflags .= try_cflags("-Wno-unused-variable");
$ccflags .= try_cflags("-Wno-gnu");
$ccflags .= try_cflags("-g");

$swig_flags = "$gsl_conf->{gsl_cflags} $swig_flags";

if ( GSLBuilder::is_cygwin()  && $Config{shrpenv} =~ m{\Aenv LD_RUN_PATH=(.*)\Z} ) {
    $ldflags .= " -L$1 -lperl";
    # Should we check the 32-ness?
    $swig_flags = '-DNEED_LONG';
} elsif (GSLBuilder::is_msys()) {
    # It is not possible to leave a symbol in a PE DLL undefined at link time, 
    # to be satisfied at runtime, as it is instead possible with most UNIX shared objects
    # See https://autotools.io/libtool/windows.html
    # So we need to link with libperl.dll.a on MSYS2
    $ldflags .= " -L$Config{archlib}/CORE -lperl";
} elsif (GSLBuilder::is_darwin()) {
    $ldflags .= ' -bundle -flat_namespace ';
}

if ($Config{archname} =~ /x86_64|amd64/ ) {
    $ldflags .= ' -fPIC  -fno-omit-frame-pointer ';
    $ccflags .= ' -fPIC  -fno-omit-frame-pointer ';
}
# this causes warnings with clang + OSX
$ccflags = $Config{cccdlflags} . ' ' . $ccflags unless GSLBuilder::is_darwin();

my $ver2func = Ver2Func->new( $gsl_conf->{gsl_version} );

my $cleanup = qq{
                 xs/*_wrap*.c core *.core swig/system.i swig/renames.i
                 swig/*.o Makefile Math-GSL-* tmp* pod2ht*.tmp _build pm
                 lib/Math/GSL/[A-z]+/* blib *.so *.orig
                } . join " ", map { catfile( (qw/lib Math GSL/, "$_.pm") ) } $ver2func->subsystems;


$ldflags = '-shared ' . $ldflags unless GSLBuilder::is_darwin();

my $swig_version;

unless (GSLBuilder::is_release()) {
    $swig_version = swig_version();
    if ((split(/\./, $swig_version))[0] < 2) {
        print "
***
*** You need to have SWIG 2.0 or greater installed. (You have $swig_version).
*** Get SWIG at http://www.swig.org/\n";
        exit 0;
    }
}

$ccflags = "$ccflags " . ($ENV{CC_FLAGS}||''),

my $builder = GSLBuilder->new(
    module_name         => 'Math::GSL',
    add_to_cleanup      => [ $cleanup ],
    dist_abstract       => 'Interface to the GNU Scientific Library using SWIG',
    dist_author         => 'Jonathan Leto <jonathan@leto.net>',
    dist_version_from   => catfile(qw/lib Math GSL.pm/),
    include_dirs        => [],
    extra_linker_flags  => $ldflags,
    extra_compiler_flags=> $ccflags . " " . $swig_flags,
    # suppress swig warnings about GSL source code
    swig_flags          => $swig_flags . " -w451,454",
    license             => 'gpl',
    meta_merge     => {
        resources => {
            repository  => 'http://github.com/leto/math--gsl/tree/master'
        },
    },
    build_requires => {
        'Test::Most'      => '0.31',
        'Test::More'      => 0,
        'Test::Exception' => '0.21',
        'Test::Class'     => '0.12',
        'Test::Taint'     => '1.06',
    },
    requires => {
        'Math::Complex'   => 0,
        'Scalar::Util'    => 0,
        'version'         => '0.77',
        'perl'            => '5.8.1',
        'parent'          => 0,
    },
    sign                  => 0,
    configure_requires => {
        'PkgConfig'       => '0.07720',
        'Module::Build'   => '0.38',
        'Alien::GSL'      => '1.01',
    },
    swig_version          => $swig_version,
    current_version       => $gsl_conf->{gsl_version},
    gsl_prefix            => $gsl_conf->{gsl_prefix},
);


$builder->add_build_element('swig');

$builder->create_build_script();

print "Have a great day!\n";


### aux functions

sub swig_version {
    my $swig = GSLBuilder::swig_binary_name();
    my $cmd  = "$swig -version";
    my $out  = `$cmd`;

    if ($?) {
        die "Can't run [$cmd] : $out";
    }
    if ($out =~ /swig\s+version\s+([\d\.]+)\n/i) {
        return $1;
    } else {
        die "Can't parse SWIG version from this output:\n$out";
    }
}

sub modify_env {
    $ENV{PKG_CONFIG_PATH} ||= '';

    # TODO: What about windows and darwin?
    my @guess = qw(/usr/pkgconfig /usr/lib/pkgconfig /usr/local/lib/pkgconfig
                   /usr/local/pkgconfig /usr/libdata/pkgconfig
                   /usr/local/libdata/pkgconfig /opt/pkgconfig);

    unshift @guess, $ENV{PKG_CONFIG_PATH} if $ENV{PKG_CONFIG_PATH};

    $ENV{PKG_CONFIG_PATH} = join(":", @guess);
}

sub try_compile {
    my ($c, %args) = @_;

    my $ok = 0;
    my $tmp = "tmp$$";
    local(*TMPC);

    my $obj_ext = $Config{_o} || ".o";
    unlink("$tmp.c", "$tmp$obj_ext");

    if (open(TMPC, ">", "$tmp.c")) {
    print TMPC $c;
    close(TMPC);

    my $cccmd = $args{cccmd};
    my $errornull;
    my $ccflags =  $Config{'ccflags'};
    $ccflags .= " $args{ccflags}" if $args{ccflags};

    if ($args{silent} ) {
        $errornull = "2>/dev/null" unless defined $errornull;
    } else {
        $errornull = '';
    }

        $cccmd = "$Config{'cc'} -o $tmp $ccflags $tmp.c $errornull"
        unless defined $cccmd;

    printf "cccmd = $cccmd\n" if $args{verbose};
    my $res = system($cccmd);
    $ok = defined($res) && $res == 0;

    if ( !$ok ) {
        my $errno = $? >> 8;
        local $! = $errno;
        print "

*** The test compile of '$tmp.c' failed: status $?
*** (the status means: errno = $errno or '$!')
*** DO NOT PANIC: this just means that you may get some innocuous
*** compiler warnings.
";
    }
    unlink("$tmp.c");

    }
    return $ok;
}

sub try_cflags {
    my ($ccflags) = @_;
    my $c_prog = "int main () { return 0; }\n";
    print "Checking if $Config{cc} supports \"$ccflags\"...";
    my $result = try_compile($c_prog, ccflags=>$ccflags);
    if ($result) {
        print "yes\n";
        return " $ccflags";
    }
    print "no\n";
    return '';
}

sub check_alien_gsl_share_dir {
    if (Alien::GSL->install_type eq "share") {
        my $dist_dir = Alien::GSL->dist_dir;
        my $bin_dir = File::Spec->catfile( $dist_dir, "bin" );
        $ENV{PATH} .= ":$bin_dir";
    }
}

sub check_gsl_version_pkgconfig {

    eval { require PkgConfig };
    if ($@) {
        print "\nI see that you are a CPANTester, you really should install PkgConfig !\n"
            if $ENV{AUTOMATED_TESTING};
        print "\nPkgConfig is currently needed to find GSL for the compilation of this module.\n";
        exit 0;
    }

    modify_env();
    print "\nAsking PkgConfig with ENV{PKG_CONFIG_PATH}=" . ($ENV{PKG_CONFIG_PATH} || '') . "\n\n";
    my $gsl_pkgcfg = PkgConfig->find ('gsl');

    if ($gsl_pkgcfg->errmsg) {
        my $errmsg = $gsl_pkgcfg->errmsg;
        print "
***
*** PkgConfig failed with error message: $errmsg
*** Probably you need to install GSL?
***  On Debian/Ubuntu you can use:
***      sudo apt-get install libgsl0-dev
***  On Mac you can use homebrew (http://brew.sh/):
***      brew install gsl
***  On Fedora/CentOS/RedHat/openSUSE you can use
***      sudo yum install gsl-devel
*** Or get GSL at http://www.gnu.org/software/gsl\n";
        exit 0;
    }

    my $gsl_version = $gsl_pkgcfg->pkg_version;
    if (GSLBuilder::cmp_versions($gsl_version, $MIN_GSL_VERSION) == -1) {
                printf "
***
*** You need to have GSL %s or greater installed. (You have $gsl_version).
*** Get GSL at http://www.gnu.org/software/gsl\n", $MIN_GSL_VERSION;
                exit 0;
            }

    return {
        gsl_version => $gsl_version,
        gsl_prefix  => $gsl_pkgcfg->get_var("prefix"),
        gsl_libs    => join (" ", $gsl_pkgcfg->get_ldflags),
        gsl_cflags  => join (" ", $gsl_pkgcfg->get_cflags),
    };
}

sub check_gsl_version {

    print "Checking for GSL using gsl-config\n";

    my $gsl_version = qx{gsl-config --version};
    if (not defined $gsl_version) {
        print "
***
*** Can't find GSL with gsl-config.
*** Trying with PkgConfig.
    ";
        return undef;
    }

    chomp($gsl_version);
    chomp(my $gsl_prefix  = qx{gsl-config --prefix});
    chomp(my $gsl_cflags  = qx{gsl-config --cflags});
    chomp(my $gsl_libs    = qx{gsl-config --libs});

    my $path_system = catfile(qw/swig system.i/);

    open my $fh, ">", "$path_system" or die "Could not create $path_system : $!";

    my $current_version;

    if ($gsl_version =~ m{\A(\d+(?:\.\d+)+)}) {
        $current_version = $1;
        my @current = split /\./, $current_version;
        print $fh "#define GSL_VERSION $current[0].$current[1]\n";
        print $fh "#define GSL_MAJOR_VERSION $current[0]\n";
        print $fh "#define GSL_MINOR_VERSION $current[1]\n";

        if (GSLBuilder::cmp_versions($current_version, $MIN_GSL_VERSION) == -1) {
            printf "
***
*** You need to have GSL %s or greater installed. (You have $gsl_version).
*** Get GSL at http://www.gnu.org/software/gsl\n", $MIN_GSL_VERSION;
            exit 0;
        } else {
            print "Found GSL $gsl_version (via gsl-config) installed in $gsl_prefix\n";
            return {
                gsl_version => $current_version,
                gsl_prefix  => $gsl_prefix,
                gsl_libs    => $gsl_libs,
                gsl_cflags  => $gsl_cflags,
            };
        }
    } else {
        print "
***
*** Can't parse GSL version $gsl_version.
*** Trying with PkgConfig.
";
        return undef;
    }
    close $fh or die "Could not close $path_system : $!";
}
